File: internal/snapshot/snapshot.go
// Package snapshot writes and checks the full index table behind a commitment.
//
// A merkle root is a fingerprint, not a copy: HintRegistry stores 32 bytes for an
// epoch of half a million leaves, and nothing recovers the leaves from it. So an
// epoch whose data lives only in the publisher's Postgres is not recoverable when
// that database is — which makes the commitment auditable by whoever already has
// the answer, and useless to everyone else.
//
// Epoch.uri exists for exactly this ("Pointer to the full index table") and
// IndexPublished emits it, so the pointer is already enumerable from chain logs.
// This package supplies what it should point at.
//
// The format is NDJSON: one header, then the coverage rows, then the leaves, in
// tree order. It is a stream on both ends because the leaf table runs to tens of
// megabytes, and a recovery that needs the whole thing in memory is one that fails
// when it is needed most.
//
// Nothing here has to be trusted. Verify rebuilds both trees from the document and
// hands back the roots; a caller compares them with what the registry holds. That
// is the whole security argument: the host is an availability question, never an
// integrity one, so a snapshot may be mirrored anywhere by anyone.
package snapshot

import (
	"bufio"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"strings"

	"github.com/ethereum/go-ethereum/common"
	"github.com/ethereum/go-ethereum/crypto"

	"github.com/Skanislav/evm-scan/internal/merkle"
)

// Format is the value of the header's "format" field. The version is part of the
// document because a restore reads files written by builds it has never seen.
const Format = "evmscan-snapshot/1"

// Header is the first line of a snapshot.
type Header struct {
	Format string `json:"format"`
	// ChainID is the chain the index is *about*, which is not necessarily the chain
	// the registry sits on; RegistryChainID is that one. Confusing the two is the
	// standing bug in this codebase, so a snapshot states both.
	ChainID         uint64 `json:"chain_id"`
	FromBlock       uint64 `json:"from_block"`
	ToBlock         uint64 `json:"to_block"`
	Root            string `json:"root"`
	CoverageRoot    string `json:"coverage_root"`
	LeafCount       int64  `json:"leaf_count"`
	AssetCount      int64  `json:"asset_count"`
	OnchainEpochID  *int64 `json:"onchain_epoch_id,omitempty"`
	Registry        string `json:"registry,omitempty"`
	RegistryChainID uint64 `json:"registry_chain_id,omitempty"`
}

// Coverage is one asset and the block range the publisher stood behind for it.
// Together these rows are the asset set, which is the part of an index that cannot
// be recovered from the chain any other way: an asset promoted locally never
// appears in listAssets, and assetKey is a keccak that does not invert.
type Coverage struct {
	Asset     common.Address `json:"asset"`
	FromBlock uint64         `json:"from_block"`
	ToBlock   uint64         `json:"to_block"`
}

// Leaf is one account and the contracts it touched, in the order the tree was
// built from.
type Leaf struct {
	Account common.Address   `json:"account"`
	Assets  []common.Address `json:"assets"`
}

// line is the wire shape. A discriminator keeps the document one stream rather than
// several concatenated ones, so a reader never has to seek.
type line struct {
	T         string           `json:"t"`
	Asset     *common.Address  `json:"asset,omitempty"`
	Account   *common.Address  `json:"account,omitempty"`
	Assets    []common.Address `json:"assets,omitempty"`
	FromBlock uint64           `json:"from_block,omitempty"`
	ToBlock   uint64           `json:"to_block,omitempty"`
}

// Writer streams a snapshot out. Call WriteCoverage for every asset, then WriteLeaf
// for every leaf in tree order, then Close.
type Writer struct {
	w   *bufio.Writer
	enc *json.Encoder
}

// NewWriter writes the header and returns a writer for the body.
func NewWriter(w io.Writer, h Header) (*Writer, error) {
	h.Format = Format
	bw := bufio.NewWriterSize(w, 64<<10)
	enc := json.NewEncoder(bw)
	if err := enc.Encode(h); err != nil {
		return nil, fmt.Errorf("write header: %w", err)
	}
	return &Writer{w: bw, enc: enc}, nil
}

func (s *Writer) WriteCoverage(c Coverage) error {
	asset := c.Asset
	return s.enc.Encode(line{T: "coverage", Asset: &asset, FromBlock: c.FromBlock, ToBlock: c.ToBlock})
}

func (s *Writer) WriteLeaf(l Leaf) error {
	acct := l.Account
	return s.enc.Encode(line{T: "leaf", Account: &acct, Assets: l.Assets})
}

func (s *Writer) Close() error { return s.w.Flush() }

// Result is what rebuilding a snapshot's trees produced.
type Result struct {
	Header       Header
	Root         common.Hash
	CoverageRoot common.Hash
	Leaves       int64
	Assets       int64
}

// Matches reports whether the rebuilt roots are the ones the registry committed.
// Both must match: the index root alone would leave the asset set unchecked, and
// the asset set is what a restore rebuilds the scan state from.
func (r Result) Matches(root, coverageRoot common.Hash) bool {
	return r.Root == root && r.CoverageRoot == coverageRoot
}

// Verify reads a snapshot and rebuilds both merkle trees from it.
//
// It recomputes every leaf rather than reading any hash out of the document, so a
// document that disagrees with itself cannot pass: the leaf hash comes from the
// account and its assets, and the coverage leaf from the asset key and its range.
// A snapshot is therefore only as good as the roots a caller checks it against.
func Verify(r io.Reader, onEach func(Coverage, bool)) (Result, error) {
	dec := json.NewDecoder(bufio.NewReaderSize(r, 64<<10))

	var res Result
	if err := dec.Decode(&res.Header); err != nil {
		return res, fmt.Errorf("read header: %w", err)
	}
	if res.Header.Format != Format {
		return res, fmt.Errorf("unknown snapshot format %q, want %q", res.Header.Format, Format)
	}

	var (
		indexLeaves    []common.Hash
		coverageLeaves []common.Hash
		seenLeaf       bool
	)
	for {
		var ln line
		if err := dec.Decode(&ln); err == io.EOF {
			break
		} else if err != nil {
			return res, fmt.Errorf("row %d: %w", len(indexLeaves)+len(coverageLeaves)+1, err)
		}
		switch ln.T {
		case "coverage":
			if ln.Asset == nil {
				return res, fmt.Errorf("coverage row %d has no asset", len(coverageLeaves)+1)
			}
			// Ordering is not cosmetic: a merkle tree is built from a sequence, so
			// coverage rows interleaved with leaves would put them in the wrong tree.
			if seenLeaf {
				return res, fmt.Errorf("coverage row after a leaf row; sections are ordered")
			}
			key := AssetKey(res.Header.ChainID, *ln.Asset)
			coverageLeaves = append(coverageLeaves, merkle.CoverageLeaf(key, ln.FromBlock, ln.ToBlock))
			if onEach != nil {
				onEach(Coverage{Asset: *ln.Asset, FromBlock: ln.FromBlock, ToBlock: ln.ToBlock}, false)
			}
		case "leaf":
			if ln.Account == nil {
				return res, fmt.Errorf("leaf row %d has no account", len(indexLeaves)+1)
			}
			seenLeaf = true
			indexLeaves = append(indexLeaves,
				merkle.LeafHash(*ln.Account, res.Header.ChainID, merkle.AssetsHash(ln.Assets)))
		default:
			return res, fmt.Errorf("unknown row type %q", ln.T)
		}
	}

	res.Leaves = int64(len(indexLeaves))
	res.Assets = int64(len(coverageLeaves))
	if res.Header.LeafCount != 0 && res.Header.LeafCount != res.Leaves {
		return res, fmt.Errorf("header says %d leaves, document has %d", res.Header.LeafCount, res.Leaves)
	}
	if res.Header.AssetCount != 0 && res.Header.AssetCount != res.Assets {
		return res, fmt.Errorf("header says %d assets, document has %d", res.Header.AssetCount, res.Assets)
	}
	if len(indexLeaves) > 0 {
		res.Root = merkle.Build(indexLeaves).Root()
	}
	if len(coverageLeaves) > 0 {
		res.CoverageRoot = merkle.Build(coverageLeaves).Root()
	}
	return res, nil
}

// Read decodes a whole snapshot into memory. Verify is the streaming path and the
// one a large epoch wants; this is for a restore, which has to hold the rows it is
// about to write anyway.
func Read(r io.Reader) (Header, []Coverage, []Leaf, error) {
	dec := json.NewDecoder(bufio.NewReaderSize(r, 64<<10))

	var h Header
	if err := dec.Decode(&h); err != nil {
		return h, nil, nil, fmt.Errorf("read header: %w", err)
	}
	if h.Format != Format {
		return h, nil, nil, fmt.Errorf("unknown snapshot format %q, want %q", h.Format, Format)
	}

	var (
		cov    []Coverage
		leaves []Leaf
	)
	for {
		var ln line
		if err := dec.Decode(&ln); err == io.EOF {
			break
		} else if err != nil {
			return h, nil, nil, err
		}
		switch ln.T {
		case "coverage":
			if ln.Asset == nil {
				return h, nil, nil, fmt.Errorf("coverage row %d has no asset", len(cov)+1)
			}
			cov = append(cov, Coverage{Asset: *ln.Asset, FromBlock: ln.FromBlock, ToBlock: ln.ToBlock})
		case "leaf":
			if ln.Account == nil {
				return h, nil, nil, fmt.Errorf("leaf row %d has no account", len(leaves)+1)
			}
			leaves = append(leaves, Leaf{Account: *ln.Account, Assets: ln.Assets})
		default:
			return h, nil, nil, fmt.Errorf("unknown row type %q", ln.T)
		}
	}
	return h, cov, leaves, nil
}

// AssetKey mirrors HintRegistry.assetKey and hintreg.AssetKey. It is duplicated
// here rather than imported so that snapshot does not depend on the registry
// client, which pulls in a node connection this package has no use for.
// hintreg_test asserts the two agree.
func AssetKey(chainID uint64, token common.Address) common.Hash {
	buf := make([]byte, 0, 28)
	for i := 7; i >= 0; i-- {
		buf = append(buf, byte(chainID>>(8*uint(i))))
	}
	buf = append(buf, token[:]...)
	return keccak(buf)
}

// URI expands a commitment_uri template.
//
// {id} is the publisher's own epoch id, which is what GET /v1/epochs/{id}/snapshot
// takes, and {chain} is the indexed chain. Neither is the on-chain epoch id: that
// one is assigned by the registry during publishIndex, which is after the uri has
// already been passed to it. A reader never needs it anyway — it finds the whole
// string in the IndexPublished log.
func URI(template string, chainID uint64, epochID int64) string {
	if template == "" {
		return ""
	}
	return strings.NewReplacer(
		"{id}", fmt.Sprint(epochID),
		"{chain}", fmt.Sprint(chainID),
	).Replace(template)
}

func keccak(b []byte) common.Hash { return crypto.Keccak256Hash(b) }

// Rebuild recomputes both roots from rows already in memory, the way Verify does
// while streaming. A restore has to hold the rows anyway, and must never take a
// root from the header — that is the number an edited document would forge.
func Rebuild(h Header, coverage []Coverage, leaves []Leaf) Result {
	res := Result{Header: h, Leaves: int64(len(leaves)), Assets: int64(len(coverage))}
	if len(coverage) > 0 {
		hs := make([]common.Hash, 0, len(coverage))
		for _, c := range coverage {
			hs = append(hs, merkle.CoverageLeaf(AssetKey(h.ChainID, c.Asset), c.FromBlock, c.ToBlock))
		}
		res.CoverageRoot = merkle.Build(hs).Root()
	}
	if len(leaves) > 0 {
		hs := make([]common.Hash, 0, len(leaves))
		for _, l := range leaves {
			hs = append(hs, merkle.LeafHash(l.Account, h.ChainID, merkle.AssetsHash(l.Assets)))
		}
		res.Root = merkle.Build(hs).Root()
	}
	return res
}

// Open reads a snapshot from a path, "-" for stdin, or an http(s) URL. A recovery
// normally starts from the uri in an IndexPublished log, which is a URL; the same
// file on disk has to work when the network is what broke.
func Open(ctx context.Context, source string) (io.ReadCloser, error) {
	switch {
	case source == "-":
		return io.NopCloser(os.Stdin), nil
	case strings.HasPrefix(source, "http://"), strings.HasPrefix(source, "https://"):
		req, err := http.NewRequestWithContext(ctx, http.MethodGet, source, nil)
		if err != nil {
			return nil, err
		}
		resp, err := http.DefaultClient.Do(req)
		if err != nil {
			return nil, err
		}
		if resp.StatusCode != http.StatusOK {
			resp.Body.Close()
			return nil, fmt.Errorf("fetch %s: %s", source, resp.Status)
		}
		return resp.Body, nil
	case strings.HasPrefix(source, "ipfs://"):
		return nil, fmt.Errorf("ipfs:// is not fetched directly; pass a gateway URL for %s", source)
	default:
		return os.Open(source)
	}
}

File: mirror/src/snapshot.ts
/**
 * Reading and checking an evmscan snapshot document.
 *
 * The format is the one `internal/snapshot` writes: NDJSON, one header line, then
 * the coverage rows, then the leaves, in tree order. This is the port of
 * `snapshot.Verify` — it recomputes every leaf from the account and its assets
 * rather than reading any hash out of the document, so a document that disagrees
 * with itself cannot pass. A snapshot is therefore only as good as the roots the
 * caller checks it against, which come from the chain.
 */

import {
  assetKey,
  assetsHash,
  buildRoot,
  coverageLeaf,
  leafHash,
  normalizeHex,
  ZERO_HASH,
  type Hex,
} from "./merkle.js";

export const FORMAT = "evmscan-snapshot/1";

export interface SnapshotHeader {
  format: string;
  /**
   * The chain the index is *about*. `registryChainId` is the chain the registry
   * sits on, which is not necessarily the same one — confusing the two is the
   * standing bug in this codebase, so a snapshot states both and so do we.
   */
  chainId: bigint;
  fromBlock: bigint;
  toBlock: bigint;
  root: Hex;
  coverageRoot: Hex;
  leafCount: number;
  assetCount: number;
  onchainEpochId?: bigint;
  registry?: Hex;
  registryChainId?: bigint;
}

export interface Coverage {
  asset: Hex;
  fromBlock: bigint;
  toBlock: bigint;
}

export interface Leaf {
  account: Hex;
  assets: Hex[];
}

export interface Snapshot {
  header: SnapshotHeader;
  coverage: Coverage[];
  leaves: Leaf[];
  /** Rebuilt from the rows, never taken from the header. */
  root: Hex;
  coverageRoot: Hex;
}

/**
 * `JSON.parse`, with integer literals preserved as strings.
 *
 * This exists because `JSON.parse` is lossy for the values this format carries:
 * `chain_id`, `from_block` and `to_block` are uint64 on the wire, and anything above
 * 2^53 comes back from `JSON.parse` already rounded — by the time a guard could
 * look at it, the true value is gone. A rounded chain id hashes to a wrong leaf, so
 * the loss would show up as a root mismatch blamed on the publisher.
 *
 * So the text is rewritten before it is parsed: every bare integer token becomes a
 * quoted string, which `toBigInt` then reads exactly. The scan tracks string state
 * so that digits inside a JSON string — a hex address, a uri with a port — are left
 * alone.
 */
function parseJsonLossless(text: string): unknown {
  let out = "";
  let inString = false;
  for (let i = 0; i < text.length; i++) {
    const ch = text[i]!;
    if (inString) {
      out += ch;
      if (ch === "\\") {
        out += text[++i] ?? "";
      } else if (ch === '"') {
        inString = false;
      }
      continue;
    }
    if (ch === '"') {
      inString = true;
      out += ch;
      continue;
    }
    // A number token starts here only if we are between structural characters.
    if (ch === "-" || (ch >= "0" && ch <= "9")) {
      let j = i;
      if (text[j] === "-") j++;
      const digits = (): void => {
        while (j < text.length && text[j]! >= "0" && text[j]! <= "9") j++;
      };
      digits();
      const intEnd = j;
      // A fraction or an exponent has to be consumed with the token, not left for
      // the next iteration: emitting only the integer part and then re-quoting the
      // remaining digits produces invalid JSON, and the failure would read as a
      // corrupt document rather than an unexpected field.
      if (text[j] === ".") {
        j++;
        digits();
      }
      if (text[j] === "e" || text[j] === "E") {
        j++;
        if (text[j] === "+" || text[j] === "-") j++;
        digits();
      }
      if (j === intEnd) {
        out += `"${text.slice(i, j)}"`;
      } else {
        // A non-integer is left as-is; no field in this format is one, and
        // JSON.parse hands it back as a number that toBigInt refuses.
        out += text.slice(i, j);
      }
      i = j - 1;
      continue;
    }
    out += ch;
  }
  return JSON.parse(out);
}

function requireField(row: Record<string, unknown>, name: string, at: string): unknown {
  const v = row[name];
  if (v === undefined || v === null) throw new Error(`${at} has no ${name}`);
  return v;
}

function toBigInt(v: unknown, at: string): bigint {
  if (typeof v === "bigint") return v;
  if (typeof v === "number") {
    if (!Number.isSafeInteger(v)) throw new Error(`${at}: ${v} is not a safe integer`);
    return BigInt(v);
  }
  if (typeof v === "string" && v !== "") return BigInt(v);
  if (v === undefined || v === null) return 0n;
  throw new Error(`${at}: cannot read ${String(v)} as an integer`);
}

/**
 * Parses the header, which JSON.parse would otherwise hand back with `chain_id` as
 * a `number`. Chain ids and block numbers are uint64 on the wire; above 2^53 a
 * `number` silently rounds and the resulting leaf hash is wrong rather than absent.
 * So every integer field is read through `toBigInt`, and a value that arrived
 * already-rounded is rejected instead of hashed.
 */
export function parseHeader(line: string): SnapshotHeader {
  const raw = parseJsonLossless(line) as Record<string, unknown>;
  const format = raw["format"];
  if (format !== FORMAT) {
    throw new Error(`unknown snapshot format ${JSON.stringify(format)}, want ${FORMAT}`);
  }
  const header: SnapshotHeader = {
    format: FORMAT,
    chainId: toBigInt(requireField(raw, "chain_id", "header"), "header.chain_id"),
    fromBlock: toBigInt(raw["from_block"], "header.from_block"),
    toBlock: toBigInt(raw["to_block"], "header.to_block"),
    root: normalizeHex(String(raw["root"] ?? ZERO_HASH), 32),
    coverageRoot: normalizeHex(String(raw["coverage_root"] ?? ZERO_HASH), 32),
    leafCount: Number(toBigInt(raw["leaf_count"], "header.leaf_count")),
    assetCount: Number(toBigInt(raw["asset_count"], "header.asset_count")),
  };
  if (raw["onchain_epoch_id"] !== undefined) {
    header.onchainEpochId = toBigInt(raw["onchain_epoch_id"], "header.onchain_epoch_id");
  }
  if (raw["registry"] !== undefined) {
    header.registry = normalizeHex(String(raw["registry"]), 20);
  }
  if (raw["registry_chain_id"] !== undefined) {
    header.registryChainId = toBigInt(raw["registry_chain_id"], "header.registry_chain_id");
  }
  return header;
}

/**
 * Reads a whole snapshot and rebuilds both trees from it.
 *
 * Row ordering is not cosmetic: a merkle tree is built from a sequence, so coverage
 * rows interleaved with leaves would land in the wrong tree. The Go writer emits
 * them in sections and this refuses a document that does not.
 */
export function parseSnapshot(text: string): Snapshot {
  const lines = text.split("\n").filter((l) => l.trim() !== "");
  const first = lines[0];
  if (first === undefined) throw new Error("empty snapshot: no header");
  const header = parseHeader(first);

  const coverage: Coverage[] = [];
  const leaves: Leaf[] = [];
  const coverageLeaves: Hex[] = [];
  const indexLeaves: Hex[] = [];
  let seenLeaf = false;

  for (let i = 1; i < lines.length; i++) {
    const row = parseJsonLossless(lines[i]!) as Record<string, unknown>;
    const at = `row ${i}`;
    switch (row["t"]) {
      case "coverage": {
        if (seenLeaf) throw new Error(`${at}: coverage row after a leaf row; sections are ordered`);
        const asset = normalizeHex(String(requireField(row, "asset", at)), 20);
        const c: Coverage = {
          asset,
          fromBlock: toBigInt(row["from_block"], `${at}.from_block`),
          toBlock: toBigInt(row["to_block"], `${at}.to_block`),
        };
        coverage.push(c);
        coverageLeaves.push(
          coverageLeaf(assetKey(header.chainId, c.asset), c.fromBlock, c.toBlock),
        );
        break;
      }
      case "leaf": {
        seenLeaf = true;
        const account = normalizeHex(String(requireField(row, "account", at)), 20);
        const assets = ((row["assets"] as string[] | undefined) ?? []).map((a) =>
          normalizeHex(a, 20),
        );
        leaves.push({ account, assets });
        indexLeaves.push(leafHash(account, header.chainId, assetsHash(assets)));
        break;
      }
      default:
        throw new Error(`${at}: unknown row type ${JSON.stringify(row["t"])}`);
    }
  }

  if (header.leafCount !== 0 && header.leafCount !== leaves.length) {
    throw new Error(`header says ${header.leafCount} leaves, document has ${leaves.length}`);
  }
  if (header.assetCount !== 0 && header.assetCount !== coverage.length) {
    throw new Error(`header says ${header.assetCount} assets, document has ${coverage.length}`);
  }

  return {
    header,
    coverage,
    leaves,
    root: buildRoot(indexLeaves),
    coverageRoot: buildRoot(coverageLeaves),
  };
}

/**
 * Rebuilds both roots from rows already in memory — the shape a mirror is in, since
 * its rows come out of SQLite rather than off the wire.
 *
 * `leaves` must be in the order the publisher built the tree from, which is
 * ascending by account: `store.SnapshotIndex` selects `ORDER BY account` over a
 * `BYTEA` column, and Postgres orders those bytewise. `sortLeaves` is that order.
 */
export function rootsOf(
  chainId: bigint,
  coverage: readonly Coverage[],
  leaves: readonly Leaf[],
): { root: Hex; coverageRoot: Hex } {
  return {
    root: buildRoot(leaves.map((l) => leafHash(l.account, chainId, assetsHash(l.assets)))),
    coverageRoot: buildRoot(
      coverage.map((c) => coverageLeaf(assetKey(chainId, c.asset), c.fromBlock, c.toBlock)),
    ),
  };
}

/**
 * Puts leaves in the order the publisher's tree was built from: ascending by
 * account, compared as bytes.
 *
 * A mirror must reproduce this exactly. Sorted-pair hashing makes a proof carry no
 * direction bits, but it does not make the tree order-independent — leaves in a
 * different order give a different root, with nothing to say which one was meant.
 */
export function sortLeaves(leaves: readonly Leaf[]): Leaf[] {
  return [...leaves].sort((a, b) => {
    const x = normalizeHex(a.account, 20);
    const y = normalizeHex(b.account, 20);
    return x < y ? -1 : x > y ? 1 : 0;
  });
}
